src/lib.rs

```diff
@@ -907,7 +907,10 @@ impl Connection {
                     // Feed crypto data to the TLS state, if there's data
                     // available at the expected offset.
                     if space.crypto_stream.readable() {
-                        let buf = space.crypto_stream.recv_pop(std::usize::MAX)?;
+                        let mut buf =
+                            vec![0; space.crypto_stream.recv_size()];
```

Comment:
I'm not sure about this. `recv_size()` returns the total amount of data that is in the buffer not what is going to be read, e.g. if there are gaps you may just end up allocating a big buffer and then read a lot less. I think we should use a while loop here as well:

```rust
let mut crypto_buf: [u8; 512] = [0; 512];

while let Ok((read, _)) = space.crypto_stream.recv_pop(&mut crypto_buf) {
    let level = space.crypto_level;

     self.tls_state.provide_data(level, &buf[..read]);
}
```

Reply:
Yeah, I had a good look at this too.  Would be nice if we could detect the different continguous blocks and gaps in the RecvBuf.  Maybe some thing in the Hashmap ordering functions so it can be detected during insertion